tt_metal/hw/firmware/src/erisc.cc

```diff
@@ -54,14 +91,65 @@ void __attribute__((section("erisc_l1_code"))) ApplicationHandler(void) {
         noc_local_state_init(n);
     }
     ncrisc_noc_full_sync();
-    while (1) {
-        if (erisc_info->num_bytes == 123) {
+    while (routing_info->routing_enabled != 1) {
+        internal_::risc_context_switch();
+    }
+    uint64_t dispatch_core_noc_encoding = get_dispatch_core_noc_encoding();
+    uint64_t eth_router_noc_encoding = uint64_t(NOC_XY_ENCODING(my_x[0], my_y[0])) << 32;
+
+    volatile tt_l1_ptr uint32_t *eth_db_semaphore_addr =
+        reinterpret_cast<volatile tt_l1_ptr uint32_t *>(eth_get_semaphore(0));
+
+    static constexpr uint32_t command_start_addr = eth_l1_mem::address_map::ERISC_APP_RESERVED_BASE;
+
+    bool db_buf_switch = false;
+    while (routing_info->routing_enabled) {
+        // FD: assume that no more host -> remote writes are pending
+        if (erisc_info->launch_user_kernel == 1) {
             kernel_init();
-            break;
+        } else if (routing_info->routing_mode == EthRouterMode::FD_SRC) {
+            eth_db_acquire(eth_db_semaphore_addr, eth_router_noc_encoding);
+            if (!routing_info->routing_enabled) {
+                break;
+            }
+            noc_semaphore_inc(
+                eth_router_noc_encoding | uint32_t(eth_db_semaphore_addr), -1);  // Two's complement addition
+            noc_async_write_barrier();
+
+            db_cb_config_t *eth_db_cb_config =
+                (db_cb_config_t
+                     *)(eth_l1_mem::address_map::CQ_CONSUMER_CB_BASE + (db_buf_switch * l1_db_cb_addr_offset));
+            const db_cb_config_t *remote_db_cb_config =
+                (db_cb_config_t *)(CQ_CONSUMER_CB_BASE + (db_buf_switch * l1_db_cb_addr_offset));
+            volatile tt_l1_ptr uint32_t *command_ptr =
+                reinterpret_cast<volatile tt_l1_ptr uint32_t *>(command_start_addr);
+            uint32_t num_buffer_transfers = command_ptr[DeviceCommand::num_buffer_transfers_idx];
+            uint32_t producer_consumer_transfer_num_pages =
+                command_ptr[DeviceCommand::producer_consumer_transfer_num_pages_idx];
+            command_ptr += DeviceCommand::NUM_ENTRIES_IN_COMMAND_HEADER;
+
+            for (uint32_t i = 0; i < num_buffer_transfers; i++) {
+                const uint32_t num_pages = command_ptr[2];
+                uint32_t num_writes_completed = 0;
+                while (num_writes_completed != num_pages) {
+                    uint32_t num_to_write = min(num_pages, producer_consumer_transfer_num_pages);
+                    multicore_eth_cb_wait_front(eth_db_cb_config, num_to_write);
+                    // contains device command, maybe just send pages, and send cmd once at the start
+                    internal_::send_fd_packets();
+                    multicore_eth_cb_pop_front(
+                        eth_db_cb_config, remote_db_cb_config, dispatch_core_noc_encoding, num_to_write);
+                    num_writes_completed += num_to_write;
```

Comment:
so the cbs handle how many pages of data are being sent / received but the first packet being sent to the dst router  contains the device command as well?

Reply:
every packet is sending the full eth FD reserved L1, so it contains cmd + data + junk L1. This is because I measured the perf of sending eth packets, and my hunch is that it's more performant to send one transaction request for a larger L1 than add risc logic to offset addresses and issue smaller L1 transfers. This is something we can tweak later though